### PR TITLE
[REG-1279] Add bots to imported sample scene

### PR DESCRIPTION
**Pull request by vontell**
The imported sample scene now includes a default bot implementation that works out of the box.

<div>
    <a href="https://www.loom.com/share/cbb2e510a9e6455f8538791a8128d1cc">
      <p>Your first bot in 60 seconds - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/cbb2e510a9e6455f8538791a8128d1cc">
      
    </a>
  </div>

Some notes:
* The generated scripts are all included, but if you want the JSON for agent builder, you will need to use the menu to do that.
* A script has been added that makes it a tiny bit easier to copy any changes from a dev sample back into the SDK
* A SDK change has been made to allow bots to be found in any location within the Assets directory

One open question:
* The namespace we are defining for actions, serializers, etc... is just RegressionGames, so if I generate scripts again, the existing actions and states and serializers in my sample project conflict (e.g. the imported sample project and my project both have a class called RGActionMap within the RegressionGames namespace). @rcornwall-reg @RG-nAmKcAz this should theoretically be an easy fix? Basically we shouldn't be using the RegressionGames namespace for code that is generated and added to their project. There are three options here:
1. Before I copy the sample code back into the SDK, I change the namespace to RGThirdPersonDemo (this is what I did for now just to get it to work, not ideal)
2. Instead of RegressionGames, we use the namespace of the RGAction and RGState classes (this almost works, there are some spots where a single class has multiple deserializers for instance from potentially multiple namespaces, but we can choose the first one maybe?) - issue, though, is that not all projects will have namespaces.
3. We randomly generate a namespace for each generation (the same thing we do for bots) - simple, but also the imports into bot code will be really weird looking.
4. We grab the game name via `PlayerSettings.productName` and add `RG` to the front (i.e. for my personal sample scene (SampleScene2), it becomes `RGSampleScene2` - I personally like this option the best, and I added code to demonstrate this, but there are places I missed and also I am not addressing the agent builder part yet. I figure if this looks like a decent change, one of you could take it on (unless my code is close to having it all)

### CODE REVIEW

File: DEVELOPMENT.md
Comment (rcornwall-reg) [inline DEVELOPMENT.md:1]: Very nice to have

**Review by rcornwall-reg — CHANGES_REQUESTED** [review]
Some small notes, but overall it looks really seamless to get going from a blank project!

File: Editor/Scripts/CodeGenerators/RGCodeGenerator.cs [hunk 1]
```diff
@@ -36,7 +36,7 @@ public class RGCodeGenerator
         // so that we don't duplicate .cs files that are already included in the sample projects.
         // But while still scanning them so that we can include their States/Actions in the json.
         private static HashSet<string> _excludeDirectories = new() {
-            "ThirdPersonDemoURP"
+           //"ThirdPersonDemoURP"
```
[the remark this answers is not in the document]
Comment (vontell) [reply Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:39]: Oops yes, will remove this before merging

File: Editor/Scripts/CodeGenerators/RGCodeGenerator.cs [hunk 2]
```diff
@@ -180,9 +180,10 @@ private static void GenerateActionClasses(List<RGActionAttributeInfo> actionInfo
                 File.Delete(directoryToDelete + ".meta");
                 AssetDatabase.Refresh();
             }
-            
-            //NOTE: We may have trouble here with serialization or actionMap with sample projects
+
+            // NOTE: We may have trouble here with serialization or actionMap with sample projects
             // Much testing needed
+            Debug.Log(PlayerSettings.productName);
```
[the remark this answers is not in the document]
Comment (vontell) [reply Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:186]: Yup sorry I keep forgetting about that! Will change it.

[the remark this answers is not in the document]
Comment (vontell) [reply Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:186]: Or in this case... this was for testing so I can remove it

File: Samples~/ThirdPersonDemoURP/Demo/Scripts/Player/PlayerInputControl.cs [hunk 3]
```diff
@@ -12,17 +13,23 @@ public class PlayerInputControl : MonoBehaviour
...
+            }
         }
 
+        [RGAction("MoveInDirection")]
         public void MoveInput(Vector2 newMoveDirection)
         {
             move = newMoveDirection;
         }
-
+
```
Comment (rcornwall-reg) [inline Samples~/ThirdPersonDemoURP/Demo/Scripts/Player/PlayerInputControl.cs:32]: Just a note that some of our IDEs add white space, which gets checked in, then others remove it, which gets checked in. Rider can share a `.DotSettings` file that would keep all of our IDEs in-sync for style and remove back-and-forth changes for things like new-lines, whitespace, etc

https://blog.jetbrains.com/dotnet/2017/02/20/rider-eap-update-code-style-settings-layers/

**Review by rcornwall-reg — APPROVED** [review]

**Comment by vontell** [comment]
Comment from Zack - there should only be one action map, those should not be committed with the sample project. The change required will be to remove those files, and then have generate scripts called when I import the scene.

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Dev Tooling
+
+## Saving a new sample from a development environment
+
+In order to make changes to the sample scenes within the SDK, you need to create a new Unity project and
+import the samples into your project. We provide a script that allows you to save your completed changes
+back into the SDK.
```
Comment (abeizer) [inline DEVELOPMENT.md:7]: I'm not sure what "saving back into the SDK" means

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+Right now, we only support one sample (ThirdPersonDemoURP), but over time more will be added - right now,
+this code only supports this sample.
```
Comment (abeizer) [inline DEVELOPMENT.md:10]: Redundant statement

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+Run the following command and follow the prompts to save your changes.
```
Comment (abeizer) [inline DEVELOPMENT.md:12]: Can probably remove the sentence above about "We provide a script that allows you to save..." since this comes immediately afterwards.

File: Samples~/ThirdPersonDemoURP/Bots/NewRGBot_n1024122209000000000/BotEntryPoint.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
+using RegressionGames;
+using RegressionGames.RGBotLocalRuntime;
+using RGThirdPersonDemo;
+using UnityEngine;
+using Random = System.Random;
+
+namespace NewRGBot_n1024122209000000000
```
Comment (abeizer) [inline Samples~/ThirdPersonDemoURP/Bots/NewRGBot_n1024122209000000000/BotEntryPoint.cs:7]: This looks identical to `QuickstartBot`. Did you mean to remove this one?

File: Samples~/ThirdPersonDemoURP/Demo/Scripts/EnemyController.cs [hunk 8]
```diff
@@ -129,6 +131,7 @@ public Sprite GetIcon()
         /*
          * Gets the enemies current hp
          */
+        [RGState("currentHealth")]
```
Comment (abeizer) [inline Samples~/ThirdPersonDemoURP/Demo/Scripts/EnemyController.cs:134]: We're mixing typecase. These states begin with a lowercase letter, but `IsAttacking` starts with an uppercase. Let's make sure they're consistent

**Review by abeizer — CHANGES_REQUESTED** [review]

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]
Capturing this as request changes so it gets resolved

there should only be one action map in the project
those generated files should not be committed with the sample project.

The change required will be to remove those files from the sample, and then have generate scripts called just after we import the scene.

**Comment by rcornwall-reg** [comment]
> Capturing this as request changes so it gets resolved
> 
> there should only be one action map in the project those generated files should not be committed with the sample project.
> 
> The change required will be to remove those files from the sample, and then have generate scripts called just after we import the scene.

I don't think it makes sense to have the action map and serialization be generated:
`Assets/RegressionGames/Runtime/GenerateScripts` and the other generated code skipped:
`Assets/ThirdPersonURP/GeneratedScripts`

I implemented this approach and found a namespace issue where, in a project called TestSample, the generated action map and serialization scripts had the namespace `RGTestSample` and the checked-in action scripts of the sample had the namespace `RGThirdPersonDemo`. The namespace of the actions isn't created during the Generate Scripts, because they're being excluded, so a compile error occurs where the action in `RGThirdPersonDemo` can't find the serializer in `RGTestSample`

TLDR: We either need to include everything in the sample (Action map and serializer), have it be fully self contained including duplicate action map and serializer, or revisit how the generated scripts of the sample interact with the project as a whole.

**Comment by rcornwall-reg** [comment]
@RG-nAmKcAz Ready review:
* I'm now filtering action map and serializers from the sample. So now _none_ of the code from the sample gets generated or included in the main project. It's fully self contained. We can talk about changing this later (i.e. slack convo) but for now it's in a consistent state

**Review by RG-nAmKcAz — APPROVED** [review]
My change concern was addressed.

Approving so that when abby's concerns are addressed this will be unblocked

File: DEVELOPMENT.md (hunk 5 above)
Comment (rcornwall-reg) [reply DEVELOPMENT.md]: Fixed

File: DEVELOPMENT.md (hunk 4 above)
Comment (rcornwall-reg) [reply DEVELOPMENT.md]: Fixed

File: DEVELOPMENT.md (hunk 6 above)
Comment (rcornwall-reg) [reply DEVELOPMENT.md:16]: Fixed

File: Samples~/ThirdPersonDemoURP/Demo/Scripts/EnemyController.cs (hunk 8 above)
Comment (rcornwall-reg) [reply Samples~/ThirdPersonDemoURP/Demo/Scripts/EnemyController.cs:134]: Fixed

File: Samples~/ThirdPersonDemoURP/Bots/NewRGBot_n1024122209000000000/BotEntryPoint.cs (hunk 7 above)
Comment (rcornwall-reg) [reply Samples~/ThirdPersonDemoURP/Bots/NewRGBot_n1024122209000000000/BotEntryPoint.cs]: Fixed/Removed

**Review by abeizer — APPROVED** [review]
